Moved uniforms and GPU shaders to new class vtkShaderProperty

drouin-simon opened the pull request:
* vtkShaderProperty stored in vtkVolume to allow sharing between mappers
* Refactored vtkUniforms with more generic code
* Added getter functions to vtkUniforms for IO purposes

drouin-simon committed:
Moved uniforms and GPU shaders to new class vtkShaderProperty
* vtkShaderProperty stored in vtkVolume to allow sharing between mappers
* Refactored vtkUniforms with more generic code
* Added getter functions to vtkUniforms for IO purposes

cpinter commented on Rendering/OpenGL2/vtkOpenGLShaderProperty.cxx:
```diff
@@ -0,0 +1,225 @@
+    }
+    ReplacementMap::iterator it = this->UserShaderReplacements.begin();
+    std::advance(it, index);
+    if( it->first.ShaderType == vtkShader::Vertex )
```
It's good practice to add curly braces even for one-liners, because when adding a second line in the if, and it's forgotten, then it will immediately cause a bug.

drouin-simon replied:
fixed

cpinter commented on Rendering/Core/vtkShaderProperty.cxx:
```diff
@@ -0,0 +1,62 @@
+
+vtkShaderProperty::vtkShaderProperty()
+{
+    this->VertexShaderCode = nullptr;
```
Please replace 4 space (or tab) intentation to 2 spaces.
(Brace indentation seems to be fine here)

drouin-simon replied:
Modified all new files to have 2 spaces.

cpinter commented on Rendering/OpenGL2/vtkOpenGLShaderProperty.h:
```diff
@@ -0,0 +1,138 @@
+     * and is mainly used by mappers when building the shaders.
+     * @return const reference to internal map holding all replacements
+     */
+    typedef std::map<const vtkShader::ReplacementSpec, vtkShader::ReplacementValue> ReplacementMap;
```
Would it make sense to add this in the vtkShaderProperty base class instead?

drouin-simon replied:
Yes, in theory. The difficulty is that the base class is part of vtkRenderingCore (so that vtkShaderProperty can be used by vtkVolume), but vtkShader and vtkShaderProgram are defined in vtkRenderingOpenGL2 and of course, vtkRenderingCore shouldn't need to know anything about vtkRenderingOpenGL2. 

Another way to go would be to split vtkShader and vtkShaderProgram into base classes that do not know anything about OpenGL and that live in vtkRenderingCore and OpenGL implementation classes (vtkOpenGLShader and vtkOpenGLShaderProgram) that live in vtkRenderingOpenGL2 but I decided not to go there.

cpinter replied:
Seems reasonable. I don't think it makes sense to generalize it too much as this will probably only be used with OpenGL I guess.

cpinter commented on Rendering/OpenGL2/vtkOpenGLUniforms.cxx:
```diff
@@ -0,0 +1,979 @@
+{
+public:
+    std::string GetGlslDeclaration( const char * name ) override
+    { return std::string("uniform int ") + name + ";\n"; }
```
Can you please indent these bodies? It makes it hard to read, at least for me.

drouin-simon replied:
fixed

cpinter commented on Rendering/OpenGL2/vtkOpenGLUniforms.cxx:
```diff
@@ -0,0 +1,979 @@
+
+    vtkOpenGLUniforms * Parent;
+
+    typedef std::map<std::string,Uniform*> UniformMapT;
```
Does the 'T' at the end have a purpose? I'd remove that.

drouin-simon replied:
Removed the T (which stands for Type, just an habit I have)

cpinter reviewed (commented):
This will be a great addition, thanks!
I added minor comments.

Can I ask what is the reason for doing the PR here and not in VTK? We need to wait until they finish the big module rework? Or to have a version accepted in our circles before sending it to the VTK guys?

drouin-simon commented:
The goal of adding these changes here instead of in the main VTK repository is to have them available as soon as possible in the Slicer nightly builds to facilitate collaboration. With the big module system rework in VTK, I suspect Slicer is not going to keep up to date with the HEAD of the main VTK repository until things are stabilized (which totally makes sense). But this means that if I want to see my changes in the nightly builds of Slicer anytime soon, they have to be merged here. 

I previously built these changes against the latest VTK (with the new module system) and it worked fine. There are only a handful of changes in CMakeLists.txt files needed to be compatible with the new module system. I can take care of it when Slicer decides to catch up with VTK again.

cpinter commented:
Thanks, Simon, looks good to me!

@jcfr @lassoan Can you please check and confirm that this can be merged?

jcfr commented:
> I suspect Slicer is not going to keep up to date with the HEAD of the main VTK repository until things are stabilized (which totally makes sense). But this means that if I want to see my changes in the nightly builds of Slicer anytime soon, they have to be merged here.

Does this mean that similar changes have been integrated in upstream VTK (or are under review) ? 

If yes, we should reference the corresponding Merge Request in the commit message.

We should avoid having major changes only leaving in Slicer/VTK fork.

drouin-simon commented:
@jcfr, I created a MR on the main VTK repository ([here](https://gitlab.kitware.com/vtk/vtk/merge_requests/5096)) and addressed all comments except for one requesting to extend the work so that it can be used in vtkPolyDataMapper, which is a pandora box I am not willing to open at this point. Then I moved here because I was in a hurry to get something working is Slicer. Since then, the code here has gone through a couple of modifications. If you can merge this PR, I'll cherry pick the commits, fix the code to use the new module system and submit a new MR to the main VTK repository.

jcfr commented:
Thanks for the follow up.

> If you can merge this PR, I'll cherry pick the commits, fix the code to use the new module system and submit a new MR to the main VTK repository.

:+1: 

I just posted more details and question at https://gitlab.kitware.com/vtk/vtk/merge_requests/5096#note_533150

> which is a pandora box I am not willing to open at this point

If there is a plan to address this, would be great to indicate this in the VTK MR. I would like to make sure the MR in the VTK upstream repo doesn't stall and is abandoned.

drouin-simon commented:
> I just posted more details and question at https://gitlab.kitware.com/vtk/vtk/merge_requests/5096#note_533150

👍
I do what is needed to get the changes merged both in Slicer/VTK and the main VTK repos. Just let me know what I need to do to make it happen as fast as possible.

I've just cherry-picked the changes on top of gitlab/VTK master so that the gitlab MR and the Slicer/VTK PR have the same code. I will push as soon as I'm done building/testing.

> If there is a plan to address this, would be great to indicate this in the VTK MR. I would like to make sure the MR in the VTK upstream repo doesn't stall and is abandoned.

I agree with Ken that it is the next logical step to integrate these shader properties with the vtkPolyDataMapper. To be honest, I haven't checked how much work it would be to make this available in vtkPolyDataMapper, and it is probably not that much, but I worry a little about the repercussions of this change, as I suspect a lot more people have been using shader replacements with vtkPolyDataMapper than with volume mappers. 

In the current state of VTK, the code to handle shader replacements is simply duplicated in all mappers. My changes just makes the implementation more elegant and easier to generalize in the volume mappers, but it does't create a special case. I suggest we merge those changes now and make sure they fit most use cases during the next couple of months and make adjustments if required. Then we can generalize to other types of mappers (who knows, maybe someone else is going to be willing to undertake this work in the meantime).

jcfr commented:
> I do what is needed to get the changes merged both in Slicer/VTK and the main VTK repos. 

:100: :+1: 

> I've just cherry-picked the changes on top of gitlab/VTK master so that the gitlab MR and the Slicer/VTK PR have the same code. I will push as soon as I'm done building/testing.

Great. As soon as this is complete and the VTK tests are  green, the Slicer/VTK fork we will update.

jcfr commented:
> I do what is needed to get the changes merged both in Slicer/VTK and the main VTK repos. 

:100: :+1: 

> I've just cherry-picked the changes on top of gitlab/VTK master so that the gitlab MR and the Slicer/VTK PR have the same code. I will push as soon as I'm done building/testing.

Great. As soon as this is complete and the VTK tests are  green, the Slicer/VTK fork we will update.

jcfr commented:
> I agree with Ken that it is the next logical step to integrate these shader properties with the vtkPolyDataMapper. [...]

When you have a chance, would be great to follow up with a similar comment in VTK MR. It will get the expected visibility and I anticipate it will help move forward.

jcfr commented:
Closing. Superseded by https://github.com/Slicer/VTK/pull/22
